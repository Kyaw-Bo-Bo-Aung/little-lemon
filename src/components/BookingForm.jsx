import React, { useState } from "react";

const BookingForm = ({ availableTimes, onDateChange, onSubmit }) => {
  const [formData, setFormData] = useState({
    date: "",
    time: "",
    guests: "1",
    occasion: "Birthday",
  });
  const [errors, setErrors] = useState();

  const handleChange = (e) => {
    const { name, value } = e.target;
    if(name === 'guests' && !Number(value)) {
      return;
    }
    setFormData((prev) => ({ ...prev, [name]: value }));

    if (name === "date") {
      onDateChange(value);
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const { date, time, guests, occasion } = formData;
    const errors = new Set();
    if(!date) {
      errors.add({
        field: "date",
        validations: [
          {
            rule: "required",
            message: "Date field is required."
          }
        ]
      })
    }
    if(!time) {
      errors.add({
        field: "time",
        validations: [
          {
            rule: "required",
            message: "Time field is required."
          }
        ]
      })
    }
    if(!guests) {
      errors.add({
        field: "guests",
        validations: [
          {
            rule: "required",
            message: "Date field is required."
          }
        ]
      })
    }
    if(!occasion) {
      errors.add({
        field: "occasion",
        validations: [
          {
            rule: "required",
            message: "Occasion field is required."
          }
        ]
      })
    }
    if(errors.size > 0) {
      setErrors(errors);
      return;
    }
    onSubmit(formData);
  };

  return (
    <>
    <h1>Book Now</h1>
      <form onSubmit={handleSubmit} className="booking-form">
        <label htmlFor="res-date">Choose date</label>
        <input
          type="date"
          id="res-date"
          name="date"
          value={formData.date}
          onChange={handleChange}
          required
        />

        <label htmlFor="res-time">Choose time</label>
        <select
          id="res-time"
          name="time"
          value={formData.time}
          onChange={handleChange}
          required
        >
          <option value="">Select a time</option>
          {availableTimes?.map((time) => (
            <option key={time} value={time}>
              {time} (Available)
            </option>
          ))}
        </select>

        <label htmlFor="guests">Number of guests</label>
        <input
          type="number"
          id="guests"
          name="guests"
          min="1"
          max="10"
          value={formData.guests}
          onChange={handleChange}
          required
        />

        <label htmlFor="occasion">Occasion</label>
        <select
          id="occasion"
          name="occasion"
          value={formData.occasion}
          onChange={handleChange}
        >
          <option>Birthday</option>
          <option>Anniversary</option>
        </select>

        <input
          type="submit"
          value="Book"
          className="submit-btn"
        />
      </form>
    </>
  );
};

export default BookingForm;
