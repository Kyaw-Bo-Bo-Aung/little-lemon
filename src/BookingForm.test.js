import { fireEvent, render, screen } from "@testing-library/react";
import BookingForm from "./components/BookingForm";

test("Renders the BookingForm heading", () => {
  render(<BookingForm />);
  const headingElement = screen.getByText("Book Now");
  expect(headingElement).toBeInTheDocument();
});

test("does add HTML validation properly", () => {
  render(<BookingForm />);
  const dateInput = screen.getByLabelText(/date/i);
  expect(dateInput).toHaveAttribute("required");
  expect(dateInput).toHaveAttribute("type", "date");
  const timeInput = screen.getByLabelText(/Choose time/i);
  expect(timeInput).toHaveAttribute("required");
});

test("allows the user to select a date", async () => {
  render(<BookingForm />);
  const dateInput = screen.getByLabelText(/date/i);
  fireEvent.change(dateInput, { target: { value: "2025-02-10" } });
  expect(dateInput).toHaveValue("2025-02-10");
});


test("renders the form correctly", () => {
    const mockOnDateChange = jest.fn(); // Mock function
    const { container } = render(<BookingForm onDateChange={mockOnDateChange} />);
    expect(container).toMatchSnapshot();
});

test("displays an error message when no date is selected", async () => {
    render(<BookingForm />);
    const submitButton = screen.getByRole("button", { name: /Book/i });
    fireEvent.click(submitButton);

    const errorMessage = await screen.findByText("Please select date.");
    expect(errorMessage).toBeInTheDocument();
  });