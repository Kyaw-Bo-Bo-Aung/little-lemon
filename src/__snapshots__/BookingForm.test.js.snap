// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the form correctly 1`] = `
<div>
  <h1>
    Book Now
  </h1>
  <form
    class="booking-form"
  >
    <label
      for="date"
    >
      Choose date
    </label>
    <input
      id="date"
      name="date"
      required=""
      type="date"
    />
    <label
      for="res-time"
    >
      Choose time
    </label>
    <select
      id="res-time"
      name="time"
      required=""
    >
      <option>
        Select a time
      </option>
    </select>
    <label
      for="guests"
    >
      Number of guests
    </label>
    <input
      name="guests"
      type="number"
    />
    <label
      for="occasion"
    >
      Occasion
    </label>
    <select
      id="occasion"
      name="occasion"
    >
      <option>
        Birthday
      </option>
      <option>
        Anniversary
      </option>
    </select>
    <input
      class="submit-btn"
      type="submit"
      value="Book"
    />
  </form>
</div>
`;
